WordPress/gutenberg
# Packages: Reimplement ESLint config as plugin

This pull request seeks to rename and repurpose the current (yet-unpublished) `eslint-config` package to an `eslint-plugin` package. This will allow the package to serve a dual-purpose of providing configurations _and_ custom rules, where the latter is not possible without being published as its own, separate ESLint plugin.

The change in usage is minimal, in regular usage amounting to:

```diff
	extends: [
-		'@wordpress/eslint-config',
+		'plugin:@wordpress/eslint-plugin/recommended',
	],
```

In the course of doing so, I had observed significant divergence, redundancy, and a lack of opportunity for granularity between the bundled rulesets. The proposed changes seek to remedy this by:

- Flattening `configs/rules` into their respective `configs/` files
   - This will also help lend clarity when proper rules are included with the plugins, separate from what are called here as "rulesets" (specific configurations of rules)
- Consolidate the default, ESNext, and ES5 rulesets
   - The default ruleset is now `esnext`, which itself is an extension of `es5`
- Separate out `jsx-a11y` and `react` configurations
   - Enables easy usage of the JavaScript-specific rules without also opting in to an assumed React code environment
   - **Note:** Technically I think the purest approach here would not define `eslint-plugin-react` and `eslint-plugin-jsx-a11y` as dependencies of `eslint-config`, but rather assume them as optional peer dependencies. For the purpose of easing developer "Getting Started" experience, I opted to leave them as dependencies, with the primary downside being that they'd be installed whilst not necessarily used by a consuming project.

**Testing Instructions:**

Verify there is no difference in linting results for the current codebase.

```
npm run lint-js
```
